masquerade.go

```diff
@@ -79,27 +83,44 @@ type Provider struct {
 	// Url used to vet masquerades for this provider
 	TestURL     string
 	Masquerades []*Masquerade
+
+	// SNIConfig has the configuration that sets if we should or not use arbitrary SNIs
+	// and which SNIs to use.
+	SNIConfig *SNIConfig
+
 	// Optional response validator used to determine whether
 	// fronting succeeded for this provider. If the validator
 	// detects a failure for a given masquerade, it is discarded.
 	// The default validator is used if nil.
 	Validator ResponseValidator
 }
 
+type SNIConfig struct {
+	UseArbitrarySNIs bool
+	ArbitrarySNIs    []string
+}
+
 // Create a Provider with the given details
-func NewProvider(hosts map[string]string, testURL string, masquerades []*Masquerade, validator ResponseValidator, passthrough []string) *Provider {
+func NewProvider(hosts map[string]string, testURL string, masquerades []*Masquerade, validator ResponseValidator, passthrough []string, sniConfig *SNIConfig) *Provider {
 	d := &Provider{
 		HostAliases:         make(map[string]string),
 		TestURL:             testURL,
 		Masquerades:         make([]*Masquerade, 0, len(masquerades)),
 		Validator:           validator,
 		PassthroughPatterns: make([]string, 0, len(passthrough)),
+		SNIConfig:           sniConfig,
 	}
 	for k, v := range hosts {
 		d.HostAliases[strings.ToLower(k)] = v
 	}
+
 	for _, m := range masquerades {
-		d.Masquerades = append(d.Masquerades, &Masquerade{Domain: m.Domain, IpAddress: m.IpAddress})
+		var sni string
+		if d.SNIConfig != nil && d.SNIConfig.UseArbitrarySNIs {
+			crc32Hash := int(crc32.ChecksumIEEE([]byte(m.IpAddress)))
+			sni = d.SNIConfig.ArbitrarySNIs[crc32Hash%len(d.SNIConfig.ArbitrarySNIs)]
+		}
+		d.Masquerades = append(d.Masquerades, &Masquerade{Domain: m.Domain, IpAddress: m.IpAddress, SNI: sni})
```

Comment:
This is great!  Could you just add a comment explaining why we're doing this?  Something like:

> Ensure that we use a consistent SNI for a given combination of IP address and SNI set.